src/inventory_ui.cpp

```diff
@@ -108,308 +393,413 @@ bool inventory_column::is_selected( const inventory_entry &entry ) const
-        }
+    if( res.entry != nullptr && !res.entry->is_selectable() ) {
+        res.entry = nullptr;
     }
 
-    return res.str();
+    return res;
 }
 
 // @todo Move it into some 'item_stack' class.
-std::vector<std::list<item *>> restack_items( const std::list<item>::const_iterator &from,
-                                              const std::list<item>::const_iterator &to )
+std::vector<std::list<item *>> restack_items( const std::list<item>::iterator &from,
+                            const std::list<item>::iterator &to )
 {
     std::vector<std::list<item *>> res;
 
     for( auto it = from; it != to; ++it ) {
         auto match = std::find_if( res.begin(), res.end(),
-            [ &it ]( const std::list<item *> &e ) {
-                return it->stacks_with( *const_cast<item *>( e.back() ) );
-            } );
+        [ &it ]( const std::list<item *> &e ) {
+            return it->stacks_with( *e.back() );
+        } );
 
         if( match != res.end() ) {
-            match->push_back( const_cast<item *>( &*it ) );
+            match->push_back( &*it );
         } else {
-            res.emplace_back( 1, const_cast<item *>( &*it ) );
+            res.emplace_back( 1, &*it );
         }
     }
 
     return res;
 }
 
-void inventory_selector::add_custom_items( const std::list<item>::const_iterator &from,
-                                           const std::list<item>::const_iterator &to,
-                                           const std::string &title,
+const item_category *inventory_selector::naturalize_category( const item_category &category, const tripoint &pos )
+{
+    const auto find_cat_by_id = [ this ]( const std::string &id ) {
+        const auto iter = std::find_if( categories.begin(), categories.end(), [ &id ]( const item_category &cat ) {
+            return cat.id == id;
+        } );
+        return iter != categories.end() ? &*iter : nullptr;
+    };
+
+    const int dist = rl_dist( p.pos(), pos );
+
+    if( dist != 0 ) {
+        const std::string suffix = direction_suffix( p.pos(), pos );
+        const std::string id = string_format( "%s_%s", category.id.c_str(), suffix.c_str() );
+
+        const auto existing = find_cat_by_id( id );
+        if( existing != nullptr ) {
+            return existing;
+        }
+
+        const std::string name = string_format( "%s %s", category.name.c_str(), suffix.c_str() );
+        const int sort_rank = category.sort_rank + dist;
+        const item_category new_category( id, name, sort_rank );
+
+        categories.push_back( std::move( new_category ) );
+    } else {
+        const auto existing = find_cat_by_id( category.id );
+        if( existing != nullptr ) {
+            return existing;
+        }
+
+        categories.push_back( category );
+    }
+
+    return &categories.back();
+}
+
+void inventory_selector::add_item( inventory_column &target_column,
+                                   const std::shared_ptr<item_location> &location,
```

Comment:
It could help, but it looks like an unsafe workaround over the deleted copy constructor (even uglier than shared pointers). `item_location` is made non-copyable on purpose: de-facto it owns the item (can dispose it on demand) and there should be only one object that can do it.